### DISCUSSION TITLE
Testing Framework

### DISCUSSION THREAD

**Discussion by jsanda** [k8ssandra/k8ssandra#540] [General]
# Overview
There is an effort underway to address the lack of test automation in the project. There are a good number of unit tests that run as part of CI. There are only a handful of integration tests that exercise k8ssandra end to end in a running Kubernetes cluster.

The choice of testing frameworks has been a hot topic as of late, particularly in the context of integration tests. Tests are currently written with [Ginkgo](https://onsi.github.io/ginkgo). There have been some discussions about using Go's [testing](https://golang.org/pkg/testing/) package instead.

# Background
The [reaper-operator](https://github.com/k8ssandra/reaper-operator) and [medusa-operator](https://github.com/k8ssandra/medusa-operator) projects were already a work in progress before k8ssandra was created. Those projects initially used the `testing`. Both of them are developed with [operator-sdk] which used `testing` in early versions prior to 1.0.  At or around 1.0, operator-sdk started using Ginkgo. 

When I upgraded reaper-operator and medusa-operator to operator-sdk 1.0, I migrated over to Ginkgo simply because that is what operator-sdk was using. I had no prior experience with Ginkgo.

When we started adding unit tests for k8ssandra, we were using Ginkgo mostly because that is what has been in use.

# Global setup/tear down
Test setup for integration tests can be expensive. K8ssandra is no exception to that. Most tests will involve a Cassandra cluster. Deploying a Cassandra cluster can easily take 3 minutes.

We can have multiple tests that perform different verifications related to that Cassandra cluster. Because it an expensive operation, we do not want to deploy the Cassandra cluster for each test. 

Code like the following would produce very slow tests:

```go
var _ = Describe("Installing k8ssandra", func() {
    Context("with a 3 node Cassandra cluster", func() {
        BeforeEach(func() {
            createCassandraCluster()
        })
        
        It("should have 3 Cassandra pods ready", func() {
            ...
        })
        
        It("should have auth enabled", func() {
            ...
        })
        
        It("should have Reaper deployed", func() {
            ...
        })
    })  
})
```

Creating a Cassandra cluster for each of these tests is simply not practical. What we need is a global setup for the container function like this:

```go
var _ = Describe("Installing k8ssandra", func() {
    Context("with a 3 node Cassandra cluster", func() {
        BeforeAll(func() {
            createCassandraCluster()
        })
        
        It("should have 3 Cassandra pods ready", func() {
            ...
        })
        
        It("should have auth enabled", func() {
            ...
        })
        
        It("should have Reaper deployed", func() {
            ...
        })
    })  
})
```

where `BeforeAll` would execute only once before all of the specs in the `Context`. Ginkgo does not currently provide anything like this,

Ginkgo lacks global setup/teardown for the containers, `Describe` and `Context`.

This Ginkgo [issue](https://github.com/onsi/ginkgo/issues/70) has a great discussion around some of these challeneges. 

It is worth noting that the [Ginkgo 2.0 proposal](https://github.com/onsi/ginkgo/issues/711) discusses adding support for `BeforeAll` and `AfterAll`. 

## Large specs
Due to the lack of global setup/teardown functions, integration tests are often combined into one large spec.

Consider the following example:

```go
var _ = Describe("Installing k8ssandra", func() {
    Context("with a 3 node Cassandra cluster", func() {
        It("should be initialized", func() {
            Expect(cassandraPod(0).IsReady()).To(BeTrue())
            
            Expect(cassandraPod(1).IsReady()).To(BeTrue())

            Expect(cassandraPod(2).IsReady()).To(BeTrue())
            
            Expect(authEnabled()).To(BeTrue())
            
            Expect(reaper.IsReady()).To(BeTrue())
        })
    })
})
```

One of the problems with large specs is the behavior of `Expect`. It fails fast. By contrast let's looks at how this might be written with `testing`:

```go
func TestInstallK8ssandra(t *testing.T) {
    if !cassandraPod(0).IsReady() {
        t.Error("cassandra pod 0 not ready")
    }
    
    if !cassandraPod(1).IsReady() {
        t.Error("cassandra pod 1 not ready")
    }
    
    if !cassandraPod(2).IsReady() {
        t.Error("cassandra pod 2 not ready")
    }
    
    if !authEnabled() {
        t.Error("auth should be enabled")
    }
    
    if !reaper.IsReady() {
        t.Error("reaper is not ready")
    }
}
```

The `Error` function logs the error message  and fails at the end of the test. This makes it much easier to perform multiple verifications without having to going through the test setup multiple times.

# Readability

One of the discussion points on which framework to use has centered readability. We want tests to be self-documenting. 

Simply using one framework over the other does not make the test code inherently more readable. This is clear from the previous examples. I think it is fair to say though that tests written with Ginkgo are inherently more verbose than those written with `testing`. This is by design.

# Table Tests
Table-driven tests are very common in Go. Ginkgo does have support for them. Here is an example from the [docs](https://onsi.github.io/ginkgo/#table-driven-tests):

```go
var _ = Describe("Math", func() {
    DescribeTable("the > inequality",
        func(x int, y int, expected bool) {
            Expect(x > y).To(Equal(expected))
        },
        Entry("x > y", 1, 0, true),
        Entry("x == y", 0, 0, false),
        Entry("x < y", 0, 1, false),
    )
})
```

Here is a corresponding version using the `testing` package:

```go
func TestGreaterThan(t *testing.T) {
    func greaterThan(x, y int) bool {
        return x > y
    }

    type tests []struct {
        name string        
        x int
        y int
        expected bool
    } {
        {name: "x > y", x: 1, y: 0, expected: true},
        {name: "x == y", x: 1, y: 1, expected: false},
        {name: "x < y", x: 0, y: 1, expected: false},
    }
    
    for _, tc := range tests {
        got := greaterThan(tc.x, tc.y)
        if got != tc.expected {
            t.Errorf("%s - expected: %t, got: %t", tc.name, tc.expected, got)
        }
    }
}
```

On the surface the Ginkgo version may look simpler. It is actually a lot more involved. Here is what the docs say:

>It’s important to understand the life-cycle of the table. The table package is a thin wrapper around Ginkgo’s DSL. DescribeTable generates a single Ginkgo Describe, within this Describe each Entry generates a Ginkgo It. This all happens before the tests run (at “testing tree construction time”). The result is that the table expands into a number of Its (one for each Entry) that are subject to all of Ginkgo’s test-running semantics: Its can be randomized and parallelized across multiple nodes.

# Test execution order
With Ginkgo I am not entirely clear on what is random by default. With `testing` tests are executed in random order. Being able to control execution order can be very helpful with integration tests. With subtests however, you have complete control over the execution order.

# Conclusion
This is not an exhaustive comparative analysis of Ginkgo and `testing`, nor is it intended to be. The features and differences that I have highlighted here make me believe `testing` is the better choice for end-to-end integration tests.

**Comment by adejanovski** [comment]
Thanks for the detailed post @jsanda, and getting that conversation started here!

I understand your concerns with Ginkgo and how the `testing` package addresses them.
But using Ginkgo should indeed not be driven by the fact that it's widely adopted in the community. If we're using Ginkgo, it's because we want to do BDD for our end to end tests. Which gives us the expressiveness that BDD comes with, as one of its primary goals is to better communicate on testing between technical and functional folks in a team.

The concerns you are expressing are developer ones on how a BDD framework will reduce your ability to design test code the way you want to. But you cannot neglect what it brings: better collaboration on tests from all parties involved, may they be devs or not.

Also, it allows us to describe the expected behaviors prior to tests being implemented (BDD being an evolution of TDD, oriented towards functional testing). Instead of writing a doc with the tests we want to implement, we'll be able to describe them with placeholders and tie the implementation to it.
In addition, it makes it less daunting for newcomers on the project to understand what the tests are doing, making it easier to get contributions and maintain the e2e tests (as they are supposed to be way more complex than unit tests). That's a major pro for using BDD, which easily balances the restrictions that a BDD framework forces on us.

If this is something we want to do (BDD), then Ginkgo is probably a solid choice as a widespread framework in the Go/k8s communities.
If not, then the `testing` package is probably all that we need, but it will force us to be careful about how we design our test scenarios. My past experience and readings (including Clean Code from Uncle Bob) taught me that comments should be kept small (and that devs tend to comment the implementation details rather than the intent) and code should be as self explanatory as possible. This can conflict with the use of concise function names in Go, which will make the code harder to read.

BDD would enforce readability and also provide test outputs that are easy to follow.

I would rather question the use of BDD and Ginkgo in the context of unit tests. I think it's not what BDD frameworks should be used for nor where they shine.

**Reply by jeffbanks** [reply]
Adding my input on the initial separation and scope of our tests from a comment in #536.

My experience has been that integration testing is verifying an integration between two components typically. That could be done without a live cluster and may include more mocks/spys as needed.  Given components A, B, & C.

1. Unit - testing A not the integration w/ B or C.

2. Integration - Testing A+B and B+C gives some confidence for A->C but the tests are written a bit differently in that they are testing at a lower level of integration.

3. End-to-End - E2E is more of a user experience something like input @ A -> C (verify output at end) which can be more black box in nature.

This [doc](https://dev.to/betoyanes/unit-vs-integration-vs-e2e-tests-4b2o) is a good reference for differences at these levels.


Also, this is a good reference for the same in addition to some thoughts from others on NOT using `Ginkgo`.
https://cluster-api.sigs.k8s.io/developer/testing.html

**Comment by emerkle826** [comment]
After today's discussion, it seems that Ginko may present more problems than it solves. Ultimately, I think we should go with a framework that is going to allow us to identify failing test scenarios quicker and allow tests to be run in parallel as much as possible. Having tests that maybe require a few more lines of code to setup is worth it if they can be run more easily in a development environment.

I am in favor of removing Ginko for our testing here, as well as migrating our unit tests away from Ginko. If we decide that we want to peruse a BDD testing approach, then Ginko maybe comes back into play.

**Reply by jeffbanks** [reply]
I agree, I've experienced the benefits that BDD can bring to `e2e` and `integration` level tests.  However, I'm in favor as well to move forward with refactoring and construction of k8ssandra tests without `Ginkgo`.  

Based on the discussion we had today with John regarding managing state & concurrent testing limitations currently in Ginkgo 1.x, it makes sense to take a more direct approach with Go's built-in testing capabilities.

**Comment by adejanovski** [comment]
Seems like the case against BDD/Gingko was strong enough for us to move back to using the `testing` package and ditch Gingko.
I'll refactor the integration test branch accordingly and push a new PR.
